hypershift-operator/controllers/hci/hostedclusterinfrastructure_controller.go

```diff
@@ -0,0 +1,197 @@
+type HostedClusterInfrastructureReconciler struct {
+	client.Client
+	ctx context.Context
+	log logr.Logger
+
+	InfraHandler InfraHandler
+}
+
+func (r *HostedClusterInfrastructureReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.HostedClusterInfrastructure{}).
+		WithOptions(controller.Options{
+			RateLimiter:             workqueue.NewItemExponentialFailureRateLimiter(1*time.Second, 10*time.Second),
+			MaxConcurrentReconciles: 1,
+		}).Complete(r)
+}
+
+func (r *HostedClusterInfrastructureReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	r.log = log
+	r.ctx = ctx
+
+	log.Info(fmt.Sprintf("Reconcile: %s", req))
+	defer log.Info(fmt.Sprintf("Reconcile: %s Done", req))
+
+	// Retrieve the HostedClusterInfrastructure custom resource
+	var hci hyperv1.HostedClusterInfrastructure
+	if err := r.Get(ctx, req.NamespacedName, &hci); err != nil {
+		log.Info("Resource was deleted")
+		return ctrl.Result{}, nil
+	}
+
+	_, infraStatus, iamStatus := GetInfraIamConditionStatus(&hci)
+	if infraStatus && iamStatus && hci.DeletionTimestamp.IsZero() {
+		log.Info(fmt.Sprintf("Nothing to do, Hosted Cluster Infrastructure is %s", string(hyperv1.AsExpected)))
+		return ctrl.Result{}, nil
+	}
+
+	// Check for a provider secret and exit
+	var providerSecret corev1.Secret
+	var err error
+
+	providerSecretName := hci.Spec.CloudProvider.Name
+	// Nothing to do if the cloudProvider secret name is not present
+	if providerSecretName == "" {
+		log.Info("No provider secret name, can't do a reconcile")
+		r.updateStatusConditionsOnChange(&hci, nil,
+			hyperv1.ProviderSecretConfigured,
+			metav1.ConditionFalse,
+			"No cloudProvider secret reference provided, create a cloudProvider secret in the namespace "+hci.Namespace,
+			hyperv1.NotAsExpected)
+
+		return ctrl.Result{}, nil
+	}
+
+	// Attempt to retreive the cloudProvider secret
+	err = r.Client.Get(ctx, types.NamespacedName{Namespace: hci.Namespace, Name: providerSecretName}, &providerSecret)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Could not retrieve the provider secret: %s", providerSecretName))
+		return ctrl.Result{RequeueAfter: 30 * time.Second, Requeue: true},
+			r.updateStatusConditionsOnChange(&hci, nil,
+				hyperv1.ProviderSecretConfigured,
+				metav1.ConditionFalse,
+				"The secret "+providerSecretName+" could not be retreived from namespace "+hci.Namespace,
+				hyperv1.NotAsExpected)
+	}
+	if err := r.updateStatusConditionsOnChange(&hci, nil, hyperv1.ProviderSecretConfigured, metav1.ConditionTrue, "Retreived secret "+providerSecretName, string(hyperv1.AsExpected)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Add a finalizer so the resource can't be deleted till the infrastructure is cleaned up
+	// Also set the Infra-ID if its not set.
+	if !controllerutil.ContainsFinalizer(&hci, DestroyFinalizer) {
```

Comment:
If I'm reading correctly, the way this is now addFinaliser will race with the garbage collector trying remove the CR after the finaliser is removed below, and so preventing the Resource from ever being deleted it addFinaliser wins the race.

Can we follow similar patern to the other controllers and the top of the reconciling loop ensure deletion

```
	// If deleted, clean up and return early.
	if !hci.DeletionTimestamp.IsZero() {
		if err := r.delete(ctx, hostedClusterInfrastructure, namespace); err != nil {
			return ctrl.Result{}, fmt.Errorf("failed to delete nodepool: %w", err)
		}
		// Now we can remove the finalizer.
		if controllerutil.ContainsFinalizer(hostedClusterInfrastructure, finalizer) {
			controllerutil.RemoveFinalizer(hostedClusterInfrastructure, finalizer)
			if err := r.Update(ctx, hostedClusterInfrastructure); err != nil {
				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from nodepool: %w", err)
			}
		}
		log.Info("Deleted hostedClusterInfrastructure", "name", req.NamespacedName)
		return ctrl.Result{}, nil
	}
```

Then

```
	// Ensure the hostedClusterInfrastructure has a finalizer for cleanup
	if !controllerutil.ContainsFinalizer(hostedClusterInfrastructure, finalizer) {
		controllerutil.AddFinalizer(hostedClusterInfrastructure, finalizer)
		if err := r.Update(ctx, hostedClusterInfrastructure); err != nil {
			return ctrl.Result{}, fmt.Errorf("failed to add finalizer to hostedClusterInfrastructure: %w", err)
		}
	}
```

Reply:
Applied this approach.
1. I moved the AddFinalizer code below the deletionTimestamp.  This should have been this way to avoid the race condition you mentioned.
2. Added the delete HCI to the destroy flow.  I did not include the contains finalizer check, as the removefinalizer does the same loop through the finalizers and the patch does nothing if the finalizer isn't removed  (this path is not to costly as it only runs once per HCI)